Add writeWord writeBlock

edspark opened the pull request:
This adds two new functions that target the same functionality as `writeByte()`: `writeWord()` and `writeBlock()`. The intent is to be able to send data over I2C without indexing to a register. This is helpful in the case where the device simply expects I2C commands over I2C. Examples of this include products like the new Ultrasonic Distance Sensor, Quad Relay, Qwiic Twist, Qwiic RFID, etc. Basically devices, at least in the examples I listed, that have micro-controllers that act as peripherals.

sfe-SparkFro commented:
Haven't reviewed, but in testing the Ultrasonic Arduino lib, I get the following warning:

```
sfeTkArdI2C.cpp:129:9: error: unused variable 'nData' [-Werror=unused-variable]
  129 |     int nData = 0;
      |         ^~~~~
```

I strongly recommend setting "Compiler warnings" in Arduino to "All" so these warnings show up

edspark commented:
I've been using the arduino-cli and didn't even think to increase verbosity! I will do that!

gigapod commented on src/sfeTk/sfeTkIBus.h:
```diff
@@ -87,14 +87,32 @@ class sfeTkIBus
   public:
     /**--------------------------------------------------------------------------
-     *  @brief Write a single byte to the device*
+     *  @brief Send a single byte to the device*
```
Since all the methods use a "write*" naming pattern, shouldn't this be Write? 

FWIW: Normally send/receive are used with network based operations - at least in my experience ....

edspark replied:
You're completely right. I was trying to get at the fact that it doesn't physically write bits/bytes to a register and so "send" seemed to fit. However, you're write. I'll just clarify that we're "Write a single byte, without indexing to a register". Something to that effect. I also have a few more things that I want to add to this pull request.

edspark commented:
I added a few more methods for reading from a device without a register parameter. This compliments the previously added write methods.  I have enabled warnings to makes sure I didn't miss those, and I've also adjusted the language around the description of the new and existing methods in the pull request. Many apologies for the disjointed manner of this pull request.

edspark commented:
Can I get another look at this? There's a library that is dependent on these functions going live this week. Thanks!

gigapod reviewed (changes requested):
Summary:
- Unclear why the read*() methods were added - they are the same as readRegister*() in the arduino impl
- Added a method  to implement writeBlock/Region() and reuse some code 
- Was the SPI implementation provided for the added methods to the bus interface?
      - And verify these make sense for an SPI device - if not, then these methods should not be in the core interface. 
- Has your changes been tested using the test action in the repo (or just compile the test sketch locally with different platforms)?

Happy to review this with you tomorrow.

edspark commented on src/sfeTkArdI2C.cpp:
```diff
@@ -120,13 +122,12 @@ sfeTkError_t sfeTkArdI2C::writeWord(uint16_t dataToWrite)
 sfeTkError_t sfeTkArdI2C::writeBlock(const uint8_t *data, size_t length)
 {
-    int nData = 0; 
     if (!_i2cPort)
```
Refactored

edspark commented on src/sfeTkArdI2C.h:
```diff
@@ -186,6 +186,45 @@ class sfeTkArdI2C : public sfeTkII2C
     sfeTkError_t writeRegister16Region(uint16_t devReg, const uint8_t *data, size_t length);
 
+    /**
+        @brief Reads a byte of data from the device.
```
removed

edspark commented on src/sfeTk/sfeTkIBus.h:
```diff
@@ -87,14 +87,32 @@ class sfeTkIBus
+     *  @retval sfeTkError_t -  kSTkErrOk on successful execution.
+     *
+     */
+    virtual sfeTkError_t writeWord(uint16_t data) = 0;
```
I ran these locally

edspark commented on src/sfeTkArdI2C.cpp:
```diff
@@ -218,6 +219,121 @@ sfeTkError_t sfeTkArdI2C::writeRegister16Region(uint16_t devReg, const uint8_t *
+//
+// Returns true on success, false on failure
+//
+sfeTkError_t sfeTkArdI2C::readByte(uint8_t dataToWrite, uint8_t &dataToRead)
```
Removed

edspark commented on src/sfeTkArdI2C.cpp:
```diff
@@ -218,6 +219,121 @@ sfeTkError_t sfeTkArdI2C::writeRegister16Region(uint16_t devReg, const uint8_t *
+//
+// Returns true on success, false on failure
+//
+sfeTkError_t sfeTkArdI2C::readWord(uint8_t dataToWrite, uint16_t &dataToRead)
```
Removed

edspark commented on src/sfeTkArdI2C.cpp:
```diff
@@ -218,6 +219,121 @@ sfeTkError_t sfeTkArdI2C::writeRegister16Region(uint16_t devReg, const uint8_t *
+sfeTkError_t sfeTkArdI2C::readBlock(uint8_t dataToWrite, uint8_t *data, size_t numBytes, size_t &readBytes)
+{
+
+    // got port
```
Removed

edspark commented on src/sfeTkArdI2C.cpp:
```diff
@@ -120,13 +122,12 @@ sfeTkError_t sfeTkArdI2C::writeWord(uint16_t dataToWrite)
 //---------------------------------------------------------------------------------
 // writeBlock()
```
Changed to `writeRegion()`

edspark commented on src/sfeTk/sfeTkIBus.h:
```diff
@@ -124,6 +124,45 @@ class sfeTkIBus
      */
     virtual sfeTkError_t writeRegisterByte(uint8_t devReg, uint8_t data) = 0;
 
+    /**
```
Agreed

edspark commented:
Issues addressed, may I get another look @gigapod? Thank you!

gigapod reviewed (approved):
Everything looks good - based on the comments and code from yesterday. 

On thing to note - Make sure this is tested with device drivers that are already using the toolkit ... it should, but this is a good regression test.
